EventFilter/GEMRawToDigi/plugins/GEMRawToDigiModule.cc

```diff
@@ -0,0 +1,195 @@
+  produces<GEMDigiCollection>(); 
+  unpackStatusDigis_ = pset.getParameter<bool>("UnpackStatusDigis");
+  if (unpackStatusDigis_){
+    produces<GEMVfatStatusDigiCollection>("vfatStatus");
+    produces<GEMGEBStatusDigiCollection>("GEBStatus");
+  }
+}
+
+void GEMRawToDigiModule::fillDescriptions(edm::ConfigurationDescriptions & descriptions)
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("InputLabel", edm::InputTag("rawDataCollector")); 
+  desc.add<bool>("useDBEMap", false); 
+}
+
+void GEMRawToDigiModule::beginRun(edm::Run const&, edm::EventSetup const& iSetup)
+{
+  if (useDBEMap_){
+    edm::ESHandle<GEMEMap> gemEMap;
+    iSetup.get<GEMEMapRcd>().get(gemEMap);
+    m_gemEMap = gemEMap.product();
+    m_gemROMap = m_gemEMap->convert();
+  }
+  else {
+    // no eMap, using dummy
+    m_gemEMap = new GEMEMap();
+    m_gemROMap = m_gemEMap->convertDummy();
+  }
+}
+
+void GEMRawToDigiModule::produce(edm::Event & e, const edm::EventSetup & iSetup)
+{
+  auto outGEMDigis = std::make_unique<GEMDigiCollection>();
+  auto outVfatStatus = std::make_unique<GEMVfatStatusDigiCollection>();
+  auto outGEBStatus = std::make_unique<GEMGEBStatusDigiCollection>();
+  // Take raw from the event
+  edm::Handle<FEDRawDataCollection> fed_buffers;
+  e.getByToken( fed_token, fed_buffers );
+  
+  for (unsigned int id=FEDNumbering::MINGEMFEDID; id<=FEDNumbering::MINGEMFEDID; ++id){ 
+    const FEDRawData& fedData = fed_buffers->FEDData(id);
+    
+    int nWords = fedData.size()/sizeof(uint64_t);
+    
+    if (nWords<5) continue;
+    const unsigned char * data = fedData.data();
+    
+    auto amc13Event = std::make_unique<AMC13Event>();
+    
+    const uint64_t* word = reinterpret_cast<const uint64_t* >(data);
+    
+    amc13Event->setCDFHeader(*word);
+    amc13Event->setAMC13header(*(++word));
+
+    // Readout out AMC headers
+    for (unsigned short i = 0; i < amc13Event->nAMC(); ++i)
+      amc13Event->addAMCheader(*(++word));
+    
+    // Readout out AMC payloads
+    for (unsigned short i = 0; i < amc13Event->nAMC(); ++i){
+      auto amcData = std::make_unique<AMCdata>();
+      amcData->setAMCheader1(*(++word));      
+      amcData->setAMCheader2(*(++word));
+      amcData->setGEMeventHeader(*(++word));
+      uint16_t amcId = amcData->BID();
+
+      // Fill GEB
+      for (unsigned short j = 0; j < amcData->GDcount(); ++j){
+	auto gebData = std::make_unique<GEBdata>();
+	gebData->setChamberHeader(*(++word));
+	
+	unsigned int m_nvb = gebData->Vwh() / 3; // number of VFAT2 blocks. Eventually add here sanity check
+	uint16_t gebId = gebData->InputID();
+	GEMDetId gemId(-1,1,1,1,1,0); // temp ID
+	for (unsigned short k = 0; k < m_nvb; k++){
+	  auto vfatData = std::make_unique<VFATdata>();
+	  vfatData->read_fw(*(++word));
+	  vfatData->read_sw(*(++word));
+	  vfatData->read_tw(*(++word));
+	  gebData->v_add(*vfatData);
+	  
+	  uint16_t bc=vfatData->BC();
+	  //uint8_t ec=vfatData->EC();
```

Comment:
Is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant